### PR TITLE
Blog: YAML vs Python orchestrators

### PR DESCRIPTION
**Pull request by elliotgunn**

### CODE REVIEW

File: src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        import subprocess
+        subprocess.run(["python", "load.py"])
+
+    extract() >> load()
+
+etl_pipeline()
+```
+
+The Python version requires decorator patterns, operator overloading (`>>`), framework-specific imports, and the implicit convention that calling `etl_pipeline()` at module scope registers it with the scheduler. None of that serves the goal, which is: run extract, then load, on a schedule. It serves the framework.
+draf
```
Comment (akeller) [inline src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md]:
```suggestion

```

File: src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+YAML severs that connection: the workflow definition is pure configuration, and tasks are external processes the orchestrator invokes and monitors without owning. Your Python script, R model, or Java service runs in its own container, exactly as it was written. The orchestrator coordinates execution without participating in it.
+
+## Asset-centric models got the model right, not the language
+
+Asset-centric models tried to fix this from the inside. Instead of "run task A, then task B," you define the data assets (tables, models, datasets) that should exist and let the framework figure out what needs to run to produce them. The dependency graph comes from the assets rather than being manually declared.
+
+### some kind of visual would be cool here
+
+A task graph says "do this, then do that;" an asset graph says "this thing should exist and be fresh." The latter is more aligned with what the business actually cares about, which is not whether extract.py ran at 6am but whether the revenue table is current.
```
Comment (akeller) [inline src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md]:
```suggestion
A task graph says "do this, then do that;" an asset graph says "this thing should exist and be fresh." The latter is more aligned with what the business actually cares about, which is not whether `extract.py` ran at 6am but whether the revenue table is current.
```
I may have missed this in other places, but when you reference a file name, wrap it in backticks so it renders as code.

File: src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+          metadata:
+            model_layer: staging
+            owner: data_team
+```
+
+The lineage graph builds automatically as workflows execute, tracking which resources each task consumes and produces. The same mechanism works for database tables, files, datasets, and infrastructure resources like S3 buckets or VMs. Thinking in terms of desired state rather than execution steps is the right model. Requiring Python to express that model is the part that doesn't hold up.
+
+## Infrastructure and analytics already went declarative
+
+Infrastructure went through the same shift a decade ago. Terraform and Kubernetes proved that declarative configuration handles heterogeneous systems better than imperative code. You don't write a program that provisions a VM, attaches a disk, and configures networking in sequence. You declare the desired state and let the system reconcile. The infrastructure-as-code movement converged on declarative formats (e.g., HCL, YAML) for the same reason orchestration is converging now: the coordination layer benefits from being separate from the execution layer.
```
Comment (akeller) [inline src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md]: Add the IaC acronym after infrastructure-as-code so we can guide them if it pops up in other places on the website (which it does!).


```suggestion
Infrastructure went through the same shift a decade ago. Terraform and Kubernetes proved that declarative configuration handles heterogeneous systems better than imperative code. You don't write a program that provisions a VM, attaches a disk, and configures networking in sequence. You declare the desired state and let the system reconcile. The infrastructure-as-code (IaC) movement converged on declarative formats (e.g., HCL, YAML) for the same reason orchestration is converging now: the coordination layer benefits from being separate from the execution layer.
```

File: src/contents/blogs/yaml-for-workflow-orchestration/index.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+At the transformation layer, dbt made the same bet. Analytics engineers write SQL, YAML handles the metadata: dependencies, tests, documentation, scheduling. The execution layer (SQL) and the orchestration layer (YAML) are explicitly separate, and that separation is why dbt could be adopted by analytics engineers who had no interest in learning a new programming framework. 
+
+When a domain matures past the point where one team writes all the code, separating the coordination layer from the execution layer is what makes the ecosystem scale. Declarative configuration wins because it's the natural format for that coordination layer: readable by people who didn't write it, parseable without a runtime, and decoupled from whatever executes underneath.
```
Comment (akeller) [inline src/contents/blogs/yaml-for-workflow-orchestration/index.md:121]: This is a 5 ⭐ point. 👏

File: src/contents/blogs/yaml-for-workflow-orchestration/index.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+**Key-value pairs:** `key: value`. The space after the colon is mandatory.
+
+**Nesting:** Indentation creates hierarchy. Use spaces, never tabs. Two spaces per level is the convention.
+
+**Lists:** A dash followed by a space marks each item. List items can be simple strings or complex objects.
+
+**Multi-line strings:** `|` preserves line breaks (useful for shell scripts and SQL). `>` folds lines into a paragraph.
+
+**Comments:** `#` comments out the rest of the line.
```
Comment (akeller) [inline src/contents/blogs/yaml-for-workflow-orchestration/index.md:190]: You _might_ have some problems with how this renders. I'm reviewing directly in GitHub, but make sure it shows up in the preview.

File: src/contents/blogs/yaml-for-workflow-orchestration/index.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+The most common parse errors: tabs instead of spaces, missing space after colon, and unquoted special characters (`:`, `#`, `{`, `}`). YAML 1.1 also parses `NO`, `yes`, `on`, and `off` as booleans, so country codes and feature flags need quotes.
+
+For a complete reference, including boolean edge cases and Kestra-specific patterns, see the [YAML basics guide](https://kestra.io/blogs/2024-10-16-yaml-for-kestra-beginners).
+
+## Where to go from here
+
+The best way to pressure-test this argument is to read a few real workflows. If the separation of orchestration and execution is real, it should be obvious in the YAML.
```
Comment (akeller) [inline src/contents/blogs/yaml-for-workflow-orchestration/index.md:208]: Do you have any recommendations for "real workflows"? I was assuming the list of links below would point me to a few specific ones, particularly around ETL.

**Review by akeller — COMMENTED** [review]
No huge concerns here. Just a few nitpicks.

File: src/contents/blogs/yaml-for-workflow-orchestration/index.md (hunk 6 above)
Comment (elliotgunn) [reply src/contents/blogs/yaml-for-workflow-orchestration/index.md:208]: yes, added a few!

**Comment by akeller** [comment]
Reviewing this now w/ the deployed preview link!

File: src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+When Python is your orchestration language, your task code runs inside the framework's execution environment. The framework owns imports and controls execution context. If you want to run a Bash script or a Node.js function, you go through the framework's abstractions for it. The orchestration layer and the execution layer share a runtime, which means changing one touches the other.
+
+YAML severs that connection: the workflow definition is pure configuration, and tasks are external processes the orchestrator invokes and monitors without owning. Your Python script, R model, or Java service runs in its own container, exactly as it was written. The orchestrator coordinates execution without participating in it.
+
+## Asset-centric models got the model right, not the language
+
+Asset-centric models tried to fix this from the inside. Instead of "run task A, then task B," you define the data assets (tables, models, datasets) that should exist and let the framework figure out what needs to run to produce them. The dependency graph comes from the assets rather than being manually declared.
+
+### some kind of visual would be cool here
```
Comment (akeller) [inline src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md]: 😆 gotta fix this before you merge.

**Review by akeller — CHANGES_REQUESTED** [review]

File: src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md (hunk 7 above)
Comment (elliotgunn) [reply src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md]: UGH i'm sorry!!!

Comment (akeller) [reply src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md]: @elliotgunn will you add a visual here or do you just want to remove it to get this merged?

Comment (elliotgunn) [reply src/contents/blogs/2026-04-10-yaml-for-workflow-orchestration/index.md]: I have an idea but it's going to require Claude to create it. waiting for my limit to reset

File: src/contents/blogs/yaml-for-workflow-orchestration/index.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    containerImage: python:3.11
+    commands:
+      - python load.py
+
+triggers:
+  - id: schedule
+    type: io.kestra.plugin.core.trigger.Schedule
+    cron: "0 6 * * *"
+```
+
```
Comment (wrussell1999) [inline src/contents/blogs/yaml-for-workflow-orchestration/index.md:49]: Is it worth having some text between the examples to explain which is which? I am not the biggest fan of two code blocks back to back

Comment (elliotgunn) [reply src/contents/blogs/yaml-for-workflow-orchestration/index.md:49]: edited that section

**Review by wrussell1999 — APPROVED** [review]
